Develop

KazuruK opened the pull request:
DefaultExtractor 2.0 - Код отрефакторен, вместо реализаций добавлены интерфейсы, добавлены тесты DefaultExtractorTest

DefaultWordFilter 2.0 - Добавлена фильтрация регистра, вместо реализаций добавлены интерфейсы, добавлены тесты DefaultWordFilterTest

Также:
добавлен тест DefaultExtractorWithDefaultWordFilterTest
Удалены все тестовые мейны из классов
Отредактирован ListOfWordsForFiltration.txt

KazuruK committed:
DefaultExtractor 1.2
DefaultWordFilter 1.0

KazuruK committed:
DefaultExtractor 1.2
DefaultWordFilter 1.0

KazuruK committed:
DefaultExtractor 1.2 - Ничего нового
DefaultWordFilter 1.2 - Добавлен фильтр пустого элемента

KazuruK committed:
DefaultExtractor 1.2 - Ничего нового
DefaultWordFilter 1.2 - Добавлен фильтр пустого элемента

KazuruK committed:
Merge branch 'encoding' of https://github.com/alozhkin/OPD-2020 into encoding

# Conflicts:
#	src/main/java/extractor/DefaultExtractor.java
#	src/main/java/extractor/Extractor.java

KazuruK committed:
DefaultExtractor 2.0 - Код отрефакторен, вместо реализаций добавлены интерфейсы, добавлены тесты DefaultExtractorTest

DefaultWordFilter 2.0 - Добавлена фильтрация регистра, вместо реализаций добавлены интерфейсы, добавлены тесты DefaultWordFilterTest

Также:
добавлен тест DefaultExtractorWithDefaultWordFilterTest
Удалены все тестовые мейны из классов
Отредактирован ListOfWordsForFiltration.txt

alozhkin commented on src/main/java/SiteParser.java:
```diff
@@ -95,10 +97,16 @@ public void start() {
                         () -> crawler.crawl(html),
                         EXECUTOR_SERVICE);
                 crawlerFuture.thenAccept(result -> linkQueue.addAll(linkFilter.filter(result, domain)));
```
program to interfaces not implementations

alozhkin commented on src/main/java/SiteParser.java:
```diff
@@ -95,10 +97,16 @@ public void start() {
                         () -> extractor.extract(html),
                         EXECUTOR_SERVICE);
-                extractorFuture.thenAccept(result -> db.insertAll(wordFilter.filter(result), domain));
+                extractorFuture.thenAccept(result -> {
```
Фильтр должен сам разбираться со своими исключениями, а не бросать их наверх

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
 
 public class DefaultWordFilter implements WordFilter {
 
+     public static void main(String[] args) {
```
Убрать в тесты весь код, который не относится к реализации

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+
+    }
+
+   /*public static void main(String[] args) throws IOException {
```
Аналогично

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
-    public List<String> filter(Set<String> words) {
-        // TODO
-        return null;
+    public Collection<String> filter(HashSet<String> words) throws IOException {
```
HashSet -> Collection
program to interfaces not implementations

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
-        // TODO
-        return null;
+    public Collection<String> filter(HashSet<String> words) throws IOException {
+        HashSet newSet = PunctuationMarkFilter(words);
```
Опять же: HashSet -> Set
\+ raw type, используй коллекции только с <>

KazuruK replied:
Исправлено еще до коммита

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+    // Фильтр ненужных слов
+    // Кпд данного метода не определен из-за кодировок
+
+    public static void UnnecessaryWordsFilter(HashSet<String> set) throws IOException {
```
Пустые строки должны разграничивать логические части длинного метода. В данном случае лучше обойтись вообще без них.

alozhkin replied:
Все эти методы не должны быть статическими. А так понимаю, они сделаны статическими ради удобства тестирования? Лучше создавать объект с помощью аннотации @BeforeEach.

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+    // Кпд данного метода не определен из-за кодировок
+
+    public static void UnnecessaryWordsFilter(HashSet<String> set) throws IOException {
+        String fileName = "src\\main\\resources\\ListOfWordsForFiltration.txt";
```
Для обеспечения работы программы на Windows И Linux следует использовать '/' в качестве разделителя

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+    public static void UnnecessaryWordsFilter(HashSet<String> set) throws IOException {
+        String fileName = "src\\main\\resources\\ListOfWordsForFiltration.txt";
+
+        String content = Files.lines(Paths.get(fileName), StandardCharsets.UTF_8).reduce("", String::concat);
```
Блох, 2е издание, гл. 51.
Разница двух подходов наглядно
https://imgur.com/a/FYkHQqn

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+        set.removeIf(String::isEmpty);
+    }
+
+    /*Iterator<Integer> iterator = set.iterator();
```
В ветке develop не должно быть больших кусков закомменченного кода.
Мартин, гл 17, c5

KazuruK replied:
Его и не было

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+
+    //Фильтр знаков препинания
+
+    public static HashSet<String> PunctuationMarkFilter(HashSet<String> set) throws IOException {
```
названия функций начинаются с маленькой буквы

KazuruK replied:
Это все тоже пофикшено в новом коде

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+    public static HashSet<String> PunctuationMarkFilter(HashSet<String> set) throws IOException {
+        HashSet<String> newSet = new HashSet<>();
+        for (String setObj : set)
+        newSet.add(delNoDigOrLet(setObj));
```
Потерялся отступ. Я не публиковал пожеланий по стилю, к сожалению, но лучше использовать скобочки даже в однострочных if, for, while, чтобы избежать подобных ошибок и чтобы было удобно добавлять код в блок.
\+ alt + ctr + I в помощь

KazuruK replied:
Отступы все есть, потому что когда рефакторил код, сделал ctrl + a, alt + ctr + I.
Это все уже в другом коде который я залил.

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+    }
+
+    //Удаляет умляубля (äöü)
+    // String result = s.replaceAll("\\W", "");
```
?

KazuruK replied:
Весь это код лучше заменить на set.removeIf(word -> word.contains("[^a-zA-Z0-9]+"))
Вот я как раз об этом, лучше было бы использовать закомменченный вариант, но он удаляет умляуты

alozhkin replied:
Ок, но закоменченный код все равно нужно убрать

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+    private static String delNoDigOrLet (String s) {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < s.length(); i++) {
+            if (Character .isLetterOrDigit(s.charAt(i)))
```
Пробельчик

KazuruK replied:
В новом коде, который и нужно было отправить изначально

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+    // Метод удаления знаков препинания из строки
+    // Кпд данного метода не определен из-за кодировок
+
+    private static String delNoDigOrLet (String s) {
```
Название метода непонятно

alozhkin replied:
Пробельчик

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,292 @@
+
+    public static HashSet<String> PunctuationMarkFilter(HashSet<String> set) throws IOException {
+        HashSet<String> newSet = new HashSet<>();
+        for (String setObj : set)
```
Весь это код лучше заменить на set.removeIf(word -> word.contains("[^a-zA-Z0-9]+"))

KazuruK replied:
И указывать алфавиты трех языков?
Потому что к примеру твой код уберет умляуты

alozhkin commented on src/main/java/extractor/DefaultExtractor.java:
```diff
@@ -2,49 +2,181 @@
+
+        Document doc = Jsoup.parse(html.toString());
 
+        //doc = Jsoup.parse( new File("D:\\Users\\Kazuru\\Downloads\\Sas&Co Tea.html"), "windows-1251", html.getUrl().toString());
```
Про закоменченный код \+ лучше не публиковать структуру своей файловой системы на гитхаб

alozhkin commented on src/main/java/extractor/DefaultExtractor.java:
```diff
@@ -2,49 +2,181 @@
+
         System.out.println(allInfo);
+
         String[] stringsArray;
```
Можно объединить инициализацию и присвоение

alozhkin commented:
Изменения должны быть в этом пулл реквесте, не закрывай его

alozhkin commented:
Я закончил ревью, сел за документацию selenium, ем свою соломку и тут до меня доходит: тесты где?

KazuruK committed:
Merge remote-tracking branch 'origin/encoding' into develop

# Conflicts:
#	src/main/java/extractor/DefaultExtractor.java
#	src/main/java/extractor/DefaultWordFilter.java
#	src/main/java/extractor/Extractor.java
#	src/main/java/extractor/WordFilter.java
#	src/main/java/main/SiteParser.java
#	src/main/resources/ListOfWordsForFiltration.txt

alozhkin commented on src/main/java/extractor/DefaultExtractor.java:
```diff
@@ -2,49 +2,24 @@
         stringsArray = allInfo.split("\\s");
-        // System.out.println(setOfWords);
+
         return new HashSet<>(Arrays.asList(stringsArray));
```
program to interface, not implementation

KazuruK replied:
Если есть вышестоящий интерфейс, то именно он должен быть типом переменной<-
Какой переменной, если ты про то что я возвращаю HashSet, то он не дает вернуть Collection, так как это интерфейс а не реализация

alozhkin replied:
Верно, столько раз увидел HashSet set = new HashSet() в старом коммите, что автоматом среагировал.

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,77 @@
+
+    //Фильтр знаков препинания
+
+    public static Collection<String> punctuationMarkFilter(Collection<String> set) {
```
Пустая строка служит для логического разделения больших методов. В данном случае лучше обойтись совсем без них

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,77 @@
+    }
+
+    //Удаление пустого элемента
+    public static void deleteBlankLines(Collection<String> set) {
```
Все эти методы не должны быть статическими. Если они стали статическими ради удобства тестирования, то лучше создавать экземляр класса с помощью аннотации @BeforeEach

alozhkin replied:
Да и в принципе, если метод однострочный, наверное лучше просто написать его. Это дискуссионно конечно, но я бы так сделал

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,77 @@
+    }
+
+    //Удаление пустого элемента
+    public static void deleteBlankLines(Collection<String> set) {
```
Название переменных должно быть осмысленно. Названия переменных не должно содержать реализацию из-за вот таких как раз случаев.

KazuruK replied:
Название переменный ты имеешь ввиду set?
И как мне хотя бы примерно ее по другому назвать? То что вышло из setToLowerCase??
Или как?

alozhkin replied:
words? setToLowerCase тоже стоит переименовать

KazuruK replied:
Ну, тогда получится лучше все set переименовать в words?

alozhkin replied:
Да, все замечания относятся сразу ко всему коду. Я не пишу одно и то же на каждом, это долго.

alozhkin commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -1,13 +1,77 @@
+    // Метод удаления знаков препинания из строки
+    // Кпд данного метода не определен из-за кодировок
+
+    private static String delNoDigOrLet(String s) {
```
Название функции неудачное

alozhkin commented on src/main/java/extractor/WordFilter.java:
```diff
@@ -1,8 +1,10 @@
 
 public interface WordFilter {
-    List<String> filter(Set<String> words);
+    Collection<String> filter(Collection<String> words) throws IOException;
```
Фильтр не должен пробрасывать исключение наверх

alozhkin replied:
Вот здесь ты не учёт моё сообщение:
"Фильтр не должен пробрасывать исключение наверх"

alozhkin commented on src/main/java/main/SiteParser.java:
```diff
@@ -95,10 +97,16 @@ public void start() {
                         () -> extractor.extract(html),
                         EXECUTOR_SERVICE);
-                extractorFuture.thenAccept(result -> db.insertAll(wordFilter.filter(result), domain));
+                extractorFuture.thenAccept(result -> {
```
Смотри выше

alozhkin commented on src/main/resources/ListOfWordsForFiltration.txt:
```diff
@@ -0,0 +1 @@
+в к он и тут все это для надо бы еще да к как или что то на in 
```
Лучше по языкам разделить на несколько файлов

KazuruK replied:
Чем лучше? Пониманием? Это усложнит метод удаления данных слов, потому что ему придется проверять наличие в 3 файлах вместо одного.

alozhkin replied:
Это не усложнение, а необходимость. Слова не должны путаться, языки должны легко подключаться.

alozhkin replied:
Модульность это основополагающая вещь в программировании.

KazuruK replied:
Окей, мне тогда поменять способ считывания из файла? А то при моем способе, для быстродействия, считывается сразу первая стока из файла, и все. То есть все слова должны быть в строку через пробел, сделать обычным нормальным способом с переходом на новою строку?

alozhkin replied:
Лучше через новую строку. Перевести файл из одного формата в другой несложно.

alozhkin commented on src/test/java/extractor/DefaultExtractorTest.java:
```diff
@@ -0,0 +1,86 @@
+class DefaultExtractorTest {
+
+    Link l = new Link("");
+    HTML h = new HTML("<!DOCTYPE html>\n" +
```
Лучше обходиться без грубой лексики. Это проект со сторонним заказчиком, нужно оставаться формальными.

alozhkin commented on src/test/java/extractor/DefaultWordFilterTest.java:
```diff
@@ -0,0 +1,95 @@
+        DefaultWordFilter.deleteBlankLines(wordSForFiltrationSet1);
+        assertEquals(wordSForFiltrationSet2.size(), 1);
+
+        assertFalse(wordSForFiltrationSet2.size() == 0);
```
Избыточно

alozhkin commented on src/test/java/extractor/DefaultWordFilterTest.java:
```diff
@@ -0,0 +1,95 @@
+        wordSForFiltrationSet.add("404..");
+        wordSForFiltrationSet.add("Зебры:");
+        wordSForFiltrationSet.add(".");
+
```
Со строками просто ужас. По одной было бы достаточно.

KazuruK replied:
Не понимаю почему ужас, может мне просто их в момент объявления подавать?

alozhkin replied:
?
Я про пустые строчки

alozhkin commented on src/test/java/extractor/DefaultWordFilterTest.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Test
+    void punctuationMarkFilterTest() {
+
```
Вообще привычнее, когда новая строка примыкает к объявлению функции. Так пишут все в проекте, так что лучше придерживаться этого стиля.

alozhkin commented on src/test/java/extractor/DefaultWordFilterTest.java:
```diff
@@ -0,0 +1,95 @@
+        expected.add("сентября");
+
+        DefaultWordFilter.unnecessaryWordsFilter(wordSForFiltrationSet);
+        assertEquals(wordSForFiltrationSet, expected);
```
Аргументы перепутаны местами

alozhkin commented on src/test/java/extractor/DefaultExtractorWithDefaultWordFilterTest.java:
```diff
@@ -0,0 +1,76 @@
+public class DefaultExtractorWithDefaultWordFilterTest {
+
+    Link l = new Link("");
+    HTML h = new HTML("<!DOCTYPE html>\n" +
```
Вынести всё в отдеьные файлы.

KazuruK replied:
Html в отдельный текстовый?

alozhkin replied:
Ага, в отдельный html

alozhkin reviewed (changes requested):
Если это возможно, посмотреть комментарии к прошлому заливу, вдруг что-то пропустил.

KazuruK committed:
DefaultExtractor 2.0 - Ничего нового.
DefaultWordFilter 2.0 - Переработан метод unnecessaryWordsFilter

Также:
Изменена структура тестов
Изменена структура фильтруемых слов (+2 файла)
Дописаны два конструктора к util.HTML.java

alozhkin commented on src/main/java/util/HTML.java:
```diff
@@ -11,6 +16,19 @@ public HTML(String html, Link url) {
+
+    public HTML(Path path) {
+        try {
+            this.html = Files.lines(path, StandardCharsets.UTF_8)
```
Вот тут ты не учёл моё сообщение о String concat:
"Блох, Effective Java, 2е издание, гл. 51.
Разница двух подходов наглядно
https://imgur.com/a/FYkHQqn"
Сорян, что не увидел на телефоне, я особо не вчитывался.

alozhkin commented on src/main/java/main/SiteParser.java:
```diff
@@ -95,10 +97,16 @@ public void start() {
                         EXECUTOR_SERVICE);
-                extractorFuture.thenAccept(result -> db.insertAll(wordFilter.filter(result), domain));
+                extractorFuture.thenAccept(result -> {
+                    try {
```
Исправить аналогично

alozhkin commented on src/test/java/extractor/DefaultWordFilterTest.java:
```diff
@@ -0,0 +1,79 @@
+        expected.add("11");
+        expected.add("сентября");
+        Collection<String> filteredSet = new DefaultWordFilter().filter(wordSForFiltrationSet);
+        assertTrue(expected.containsAll(filteredSet));
```
Избыточная проверка. Следующая строчка покрывает её

alozhkin reviewed (changes requested):
Хорошо. Исправь последние замечания и смержу. Осталось совсем немного сделать.

KazuruK committed:
Поправлен проброс исключений;
Переписан конструктор к util.HTML.java;
